public/app/features/dashboard/dashgrid/LazyLoader.tsx

```diff
@@ -0,0 +1,60 @@
+import React, { useRef, useState } from 'react';
+import { useUniqueId } from 'app/plugins/datasource/influxdb/components/useUniqueId';
+import { useEffectOnce, useLifecycles } from 'react-use';
+
+export interface Props {
+  children: React.ReactNode | (({ isInView }: { isInView: boolean }) => React.ReactNode);
+  width?: number;
+  height?: number;
+  onLoad?: () => void;
+  onChange?: (isInView: boolean) => void;
+}
+
+export function LazyLoader({ children, width, height, onLoad, onChange }: Props) {
+  const id = useUniqueId();
+  const [loaded, setLoaded] = useState(false);
+  const [isInView, setIsInView] = useState(false);
+  const wrapperRef = useRef<HTMLDivElement>(null);
+  useEffectOnce(() => {
+    LazyLoader.addCallback(id, (entry) => {
+      if (!loaded && entry.isIntersecting) {
+        setLoaded(true);
+        onLoad?.();
+      }
+
+      setIsInView(entry.isIntersecting);
+      onChange?.(entry.isIntersecting);
+    });
+  });
+
+  useLifecycles(
+    () => {
+      if (wrapperRef.current) {
+        LazyLoader.observer.observe(wrapperRef.current);
+      }
+    },
+    () => {
+      delete LazyLoader.callbacks[id];
+      if (Object.keys(LazyLoader.callbacks).length === 0) {
+        LazyLoader.observer.disconnect();
+      }
+    }
+  );
```

Comment:
is there an advantage to using `useLifecycles` here instead of putting it in the `useEffectOnce`? had to go google how `useLifecycles` behaves, whereas `useEffectOnce` is self explanatory if you know what `useEffect` does

Reply:
Good point! `useEffectOnce` seems to work just fine. For some reason I thought it ran before the component was actually mounted and rendered, but that's not the case. Thanks for catching this 👍